core/modules/file/file.entity.inc

```diff
@@ -141,16 +141,56 @@ class File extends Entity {
     );
   }
 
+  /**
+   * Overrides Entity::createAccess().
+   */
+  public static function createAccess($bundle = NULL, $account = NULL) {
+    $rights = &backdrop_static('file_access', array());
+
+    // If no user object is supplied, the access check is for the current user.
+    if (empty($account)) {
+      $account = $GLOBALS['user'];
+    }
+
+    // If we've already checked access for this file, user and op, return from
+    // cache.
+    if (isset($rights[$account->uid][$bundle])) {
+      return $rights[$account->uid][$bundle];
```

Comment:
Why are we caching access per UID instead of role?

Reply:
This is no different than other entity classes, such as node. It has to be on the user because role is not the only way to determine access.